### PR TITLE
rapids-cmake dependency tracking now understands COMPONENTS

### LINKED ISSUE #228: [FEA] Make dependency tracking aware of COMPONENTS
**Is your feature request related to a problem? Please describe.**
Dependency tracking, i.e. exporting targets for consumption of libraries, currently is not aware of the components functionality. This makes it so that consumers need to call additional find_packages for finding/defining the component targets. 

This came up in https://github.com/rapidsai/cuml/pull/4818, where `cuML` uses the `nn` and `distance` components of RAFT, but if a consumer calls find cuML but no find RAFT is called then this happens: 

```bash
CMake Error at /home/ursa/miniconda3/envs/skbuild4/lib/cmake/cuml/cuml-targets.cmake:56 (set_target_properties):
  The link interface of target "cuml::cuml++" contains:

    raft::nn

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  /home/ursa/miniconda3/envs/skbuild4/lib/cmake/cuml/cuml-config.cmake:77 (include)
  CMakeLists.txt:68 (find_package)
```

which makes it necessary to call a find RAFT in addition to find cuML.

### PR DESCRIPTION
**Pull request by robertmaynard**
Fixes #228

### CODE REVIEW

File: rapids-cmake/find/package.cmake
Comment (vyasr) [inline rapids-cmake/find/package.cmake:122]: Why don't we need to convert semicolons to spaces again here? Won't we have the same issue as in `rapids_cpm_find`?

File: testing/cpm/cpm_find-components/CMakeLists.txt
Comment (vyasr) [inline testing/cpm/cpm_find-components/CMakeLists.txt:19]: Assuming that #227 merges first, all of the versions in these tests will need to get updated.

File: testing/cpm/cpm_find-components/fakedependency-config.cmake
Comment (vyasr) [inline testing/cpm/cpm_find-components/fakedependency-config.cmake:23]:
```suggestion
    string(APPEND _FAIL_REASON "component '${comp}' was requested, but not found. ")
```

File: testing/export/export_package-build-with-components-and-version.cmake
Comment (vyasr) [inline testing/export/export_package-build-with-components-and-version.cmake:2]: The copyright years are off on all (or at least most) of the new files in testing/export.

File: testing/export/export_package-install-with-components-and-version.cmake
Comment (vyasr) [inline testing/export/export_package-install-with-components-and-version.cmake:18]:
```suggestion
rapids_export_package( install FAKE_PACKAGE test_export_set VERSION 14.7.2 COMPONENTS abc def)
```

Comment (vyasr) [reply testing/export/export_package-install-with-components-and-version.cmake:18]: I think you need `s/build/install/g` in all of the `export_package-install*` files.

File: testing/export/export_package-install-with-version.cmake
Comment (vyasr) [inline testing/export/export_package-install-with-version.cmake:16]: This file's copyright needs to be updated.

File: testing/find/find_package-components-failed.cmake
Comment (vyasr) [inline testing/find/find_package-components-failed.cmake:21]: Maybe also toss in `INSTALL_EXPORT_SET` for completeness?

File: rapids-cmake/export/package.cmake
Comment (vyasr) [inline rapids-cmake/export/package.cmake:91]: You could probably simplify this a bit more if you reworked the logic to something like
```
if(_RAPIDS_COMPONENTS)
  # Set common things
  if (_RAPIDS_VERSION)
    # set version and some string to stick in the components filename.
  else()
    ...
  endif()
  configure_file...
```

File: rapids-cmake/export/template/build_package_components_versioned.cmake.in
Comment (vyasr) [inline rapids-cmake/export/template/build_package_components_versioned.cmake.in:10]: Is there no way to make this entire line a variable for replacement so that you could combine the versioned and non-versioned files? Same question for the `install_package*` equivalents.

**Review by vyasr — COMMENTED** [review]
There's a few things that need to be addressed here, but I've left comments rather than requesting changes since I'm out next week.

File: rapids-cmake/find/package.cmake
Comment (robertmaynard) [reply rapids-cmake/find/package.cmake:122]: We don't have the same issue as the results getting directly written to a `find_package` command. The `rapids_cpm_find` does an extra level of argument parsing which treats the semi-colons as argument separators and therefore drops everything after the first component

File: rapids-cmake/export/template/build_package_components_versioned.cmake.in
Comment (robertmaynard) [reply rapids-cmake/export/template/build_package_components_versioned.cmake.in:10]: This is a good question. I think I can unify everything.
Let me see if it makes things easier to read

Comment (robertmaynard) [reply rapids-cmake/export/template/build_package_components_versioned.cmake.in:10]: When I moved to doing this in CMake logic it became harder to see what the final output of the request would be. While the current approach has lots of file the upside is a clearer understanding of what the template looks like

Comment (vyasr) [reply rapids-cmake/export/template/build_package_components_versioned.cmake.in:10]: OK that's fine to leave as is, then.

File: testing/cpm/cpm_find-components/CMakeLists.txt
Comment (robertmaynard) [reply testing/cpm/cpm_find-components/CMakeLists.txt]: updated

**Review by vyasr — APPROVED** [review]
Sorry not sure if you were waiting on my approval to merge this. It LGTM.

**Comment by robertmaynard** [comment]
@gpucibot merge

**Pull request merged**
